File: leetcode/C++/278.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
  int firstBadVersion(int n) {
    unsigned long int i = 0;
    unsigned long int j = n;
    while (i <= j) {
      // 0 ~ 4294967295
      unsigned long int m = (i + j) / 2;

      if (isBadVersion(m) == true && isBadVersion(m - 1) == false) {
        return m;
      }

      if (isBadVersion(m) == true) {
        j = m - 1;
      } else {
        i = m + 1;
      }
    }
    return 0;
  }
};

File: leetcode/python/278.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):


class Solution(object):
    def firstBadVersion(self, n):
        """
        :type n: int
        :rtype: int
        """
        j = n
        i = 0
        while i <= j:
            m = (i + j) / 2

            isfirst = isBadVersion(m) and isBadVersion(m - 1) == False

            if isfirst:
                return m
            elif isBadVersion(m):
                j = m - 1
            else:
                i = m + 1

		# not found
		return 0
